File: interfaces.java
import java.lang.Math;

// Interfaz Figure
interface Figure {
    String getColor();
    double perimeter();
    double area();
}

// Clase para el triángulo con sus atributos concretos
class Triangle implements Figure {
    String color;
    double side1, side2, side3;

    public Triangle(String color, double side1, double side2, double side3) {
        this.color = color;
        this.side1 = side1;
        this.side2 = side2;
        this.side3 = side3;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return side1 + side2 + side3;
    }

    @Override
    public double area() {
        double s = (side1 + side2 + side3) / 2;
        return Math.sqrt(s * (s - side1) * (s - side2) * (s - side3));
    }
}

// Clase para el círculo con sus atributos concretos
class Circle implements Figure {
    String color;
    double radius;

    public Circle(String color, double radius) {
        this.color = color;
        this.radius = radius;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return 2 * Math.PI * radius;
    }

    @Override
    public double area() {
        return Math.PI * Math.pow(radius, 2);
    }
}

// Clase para el rectángulo con sus atributos concretos
class Rectangle implements Figure {
    String color;
    double length, width;

    public Rectangle(String color, double length, double width) {
        this.color = color;
        this.length = length;
        this.width = width;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return 2 * (length + width);
    }

    @Override
    public double area() {
        return length * width;
    }
}

// Clase para el hexágono con sus atributos concretos
class Hexagon implements Figure {
    String color;
    double side;

    public Hexagon(String color, double side) {
        this.color = color;
        this.side = side;
    }

    @Override
    public String getColor() {
        return color;
    }

    @Override
    public double perimeter() {
        return 6 * side;
    }

    @Override
    public double area() {
        return (3 * Math.sqrt(3) * Math.pow(side, 2)) / 2;
    }
}

// Se crean instancias de cada clase de figura geométrica, se guardan en un
// array de tipo 'Figure' gracias al polimorfismo, se hace un for each en cada figura
// se accede a cada figura y se llaman a los métodos
public class Main {
    public static void main(String[] args) {
        Triangle triangle = new Triangle("Rojo", 3, 4, 5);
        Circle circle = new Circle("Azul", 5);
        Rectangle rectangle = new Rectangle("Verde", 4, 6);
        Hexagon hexagon = new Hexagon("Amarillo", 4);

        Figure[] figures = {triangle, circle, rectangle, hexagon};

        for (Figure figure : figures) {
            System.out.println("Color: " + figure.getColor());
            System.out.println("Perímetro: " + figure.perimeter());
            System.out.println("Área: " + figure.area());
            System.out.println();
        }
    }
}

File: interfaces.py
import math
from abc import ABC, abstractmethod

# Interfaz Figure con métodos abstractos para perímetro y área
class Figure(ABC):
    def __init__(self, color):
        self.color = color

    def getColor(self):
        return self.color

    @abstractmethod
    def perimeter(self):
        pass

    @abstractmethod
    def area(self):
        pass

# Clase para el triángulo con sus respectivas fórmulas para perímetro y área
class Triangle(Figure):
    def __init__(self, color, side1, side2, side3):
        super().__init__(color)
        self.side1 = side1
        self.side2 = side2
        self.side3 = side3

    def perimeter(self):
        return self.side1 + self.side2 + self.side3

    def area(self):
        s = (self.side1 + self.side2 + self.side3) / 2
        return math.sqrt(s * (s - self.side1) * (s - self.side2) * (s - self.side3))

# Clase para el círculo con sus respectivas fórmulas para perímetro y área
class Circle(Figure):
    def __init__(self, color, radius):
        super().__init__(color)
        self.radius = radius

    def perimeter(self):
        return 2 * math.pi * self.radius

    def area(self):
        return math.pi * (self.radius ** 2)

# Clase para el rectángulo con sus respectivas fórmulas para perímetro y área
class Rectangle(Figure):
    def __init__(self, color, length, width):
        super().__init__(color)
        self.length = length
        self.width = width

    def perimeter(self):
        return 2 * (self.length + self.width)

    def area(self):
        return self.length * self.width

# Clase para el hexágono con sus respectivas fórmulas para perímetro y área
class Hexagon(Figure):
    def __init__(self, color, side):
        super().__init__(color)
        self.side = side

    def perimeter(self):
        return 6 * self.side

    def area(self):
        return (3 * math.sqrt(3) * (self.side ** 2)) / 2

# Instancias de clase derivada con los datos como el color y las dimensiones
triangle = Triangle("Rojo", 3, 4, 5)
circle = Circle("Azul", 5)
rectangle = Rectangle("Verde", 4, 6)
hexagon = Hexagon("Amarillo", 4)

# Arreglo
figures = [triangle, circle, rectangle, hexagon]

# Ejecución
for figure in figures:
    print("Color:", figure.getColor())
    print("Perímetro:", figure.perimeter())
    print("Área:", figure.area())
    print()